src/plugins/discover/public/application/main/components/top_nav/discover_topnav.tsx

```diff
@@ -185,6 +185,11 @@ export const DiscoverTopNav = ({
     ]
   );
 
+  const onEditDataView = async (editedDataView: DataView) => {
+    stateContainer.dataState.fetch();
+    stateContainer.actions.setDataView(editedDataView);
+  };
```

Comment:
We need to take care of ad-hoc ones differently

```suggestion
  const onEditDataView = async (editedDataView: DataView) => {
   if (!editedDataView.isPersisted()) {
      await updateAdHocDataViewId(editedDataView);
    } else {
      stateContainer.actions.setDataView(editedDataView);
    }
    stateContainer.dataState.fetch();
  };
```

Reply:
@kertal Off hand, do you know why the code works this way? I was a bit surprised and disappointed to see this special handling of a different kind of data view. I'm able to create and edit ad hoc data views without your code suggestion.

Reply:
ad-hoc have a special treatment in Discover. The "stateContainer.actions.setDataView" is pretty new, internal state was just a prop high up in the component tree. There's a potential to simplify in the future, this state refactoring provides the way to it

Discover behaves like Lens when editing an Ad-hoc data view ... every time an ad-hoc one is edited, a new data view id has to be generated, to omit having duplicate id's with different settings. You could jump to Lens with a given ad-hoc id to generate a visualization, save it there. In Discover you could change the same data view, create a saved search, add the visualization and the saved search on a Dashboard, they then would have the same id, with different settings. That's why there's a special handling of ad hoc data views.

It didn't work for me when I've been editing an ad-hoc one, that's why I aligned here with @dimaanj's PR